src/Raven.Server/Documents/DocumentPutAction.cs

```diff
@@ -240,6 +233,40 @@ public PutOperationResults PutDocument(DocumentsOperationContext context, string
                     }
                 }
 
+                if (document.TryGetMetadata(out BlittableJsonReaderObject docMetadata))
+                {
+                    bool shouldRebuildDocument = false;
+                    if (newFlags.Contain(DocumentFlags.Archived))
+                    {
+                        // If document has archived flag, but @archived is dropped, rebuild it
+                        // If document has archived flag, but @archived is set to false, revert it to true
+                        if (docMetadata.TryGet(Constants.Documents.Metadata.Archived, out bool isArchived) == false || isArchived == false)
+                        {
+                            docMetadata.Modifications = new DynamicJsonValue(docMetadata);
+                            docMetadata.Modifications[Constants.Documents.Metadata.Archived] = true;
+                            shouldRebuildDocument = true;
+                        }
+                    }
+                    else
+                    {
+                        // If document has no archived flag, but @archived is in metadata, drop the metadata entry
+                        if (docMetadata.TryGet(Constants.Documents.Metadata.Archived, out bool _))
+                        {
+                            docMetadata.Modifications = new DynamicJsonValue(docMetadata);
+                            docMetadata.Modifications.Remove(Constants.Documents.Metadata.Archived);
+                            shouldRebuildDocument = true;
+                        }
+                    }
+
+                    if (shouldRebuildDocument)
```

Comment:
You can check if `docMetadata.Modifications != null` here, right?

Reply:
Done, replaced